src/Main.hs

```diff
@@ -188,38 +188,93 @@ andRaw [] = BoolLit True
-
-        -- translate wildcards into negation of other cases
-        normalize = snd . mapAccumL (\a b -> case b of
-                                        Leaf pn WildExp p -> (a, Leaf pn (ENot nowhere (andRaw a)) p)
-                                        Branch pn WildExp p -> (a, Branch pn (ENot nowhere (andRaw a)) p)
-                                        e@(Leaf _ c _) -> (c:a, e)
-                                        e@(Branch _ c _) -> (c:a, e)) []
-
-        -- split case into pass and fail case
-        splitCase ifs [] ret storage postc contracts = [Behaviour name Pass False contract iface (mconcat ifs) postc contracts storage ret]
-        splitCase ifs iffs ret storage postc contracts = [ Behaviour name Pass False contract iface (mconcat (ifs <> iffs)) postc contracts storage ret
-                                                         , Behaviour name Fail False contract iface (And (mconcat ifs) (Neg (mconcat iffs))) postc contracts storage Nothing]
-
-        -- flatten case tree
-        flatten iff postc pathcond (Leaf _ cond post) = do c <- checkBool env cond
-                                                           (p, maybeReturn, contracts) <- checkPost env contract post
-                                                           return $ splitCase (c:pathcond) iff maybeReturn p (mconcat postc) contracts
-
-        flatten iff postc pathcond (Branch _ cond cs) = do c <- checkBool env cond
-                                                           leaves <- mapM (flatten iff postc (c:pathcond)) (normalize cs)
-                                                           return $ join leaves
-
-splitBehaviour store (Constructor name contract decls iffs cases post ensures invariants) = Ok [] --error "TODO: check constructor"
+  where
+    env = (fromMaybe mempty (Map.lookup contract store), store, abiVars)
+    abiVars = Map.fromList $ map (\(Decl typ var) -> (var, metaType typ)) decls
+
+    -- translate wildcards into negation of other cases
+    normalize :: [Case Expr b] -> [Case Expr b]
+    normalize =
+      snd
+        . mapAccumL
+          ( \a b -> case b of
+              Leaf pn WildExp p -> (a, Leaf pn (ENot nowhere (andRaw a)) p)
+              Branch pn WildExp p -> (a, Branch pn (ENot nowhere (andRaw a)) p)
+              e@(Leaf _ c _) -> (c : a, e)
+              e@(Branch _ c _) -> (c : a, e)
+          )
+          []
+
+    -- flatten case tree
+    flatten :: [Exp Bool] -> [Exp Bool] -> [Exp Bool] -> Case Expr Post -> Err [Behaviour]
+    flatten iff postc pathcond (Leaf _ cond post) = do
+      c <- checkBool env cond
+      (p, maybeReturn, contracts) <- checkPost env contract post
+      return $ splitCase name False contract iface (c : pathcond) iff maybeReturn p postc contracts
+    flatten iff postc pathcond (Branch _ cond cs) = do
+      c <- checkBool env cond
+      leaves <- mapM (flatten iff postc (c : pathcond)) (normalize cs)
+      return $ join leaves
```

Comment:
Most of this was just adding type signatures and  reformatted it for line width.

The only real change was to pull `splitCase` out so it could be shared between the two pattern matches.

Reply:
Couöd be nice to lift out env and abivars similarly?